framework/resource/retryresource/crud_resource_wrapper.go

```diff
@@ -0,0 +1,100 @@
+package retryresource
+
+import (
+	"context"
+
+	"github.com/cenkalti/backoff"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+
+	originalframework "github.com/giantswarm/operatorkit/framework"
+	"github.com/giantswarm/operatorkit/framework/resource/internal"
+	"github.com/giantswarm/operatorkit/framework/resource/internal/framework"
+)
+
+// crudResourceWrapper is a specialized wrapper which wraps
+// *framework.CRUDResource.
+type crudResourceWrapper struct {
+	logger   micrologger.Logger
+	resource framework.Resource
+
+	backOff backoff.BackOff
+}
+
+func newCRUDResourceWrapper(config Config) (*crudResourceWrapper, error) {
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.Logger must not be empty")
+	}
+	if config.Resource == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.Resource must not be empty")
+	}
+
+	if config.BackOff == nil {
+		config.BackOff = backoff.NewExponentialBackOff()
+	}
+
+	// Wrap underlying resource Ops with retry logic. Underlying resource
+	// is a pointer so we can modify it in place.
+	{
+		underlying, ok := internal.Underlying(config.Resource).(*originalframework.CRUDResource)
+		if !ok {
+			return nil, microerror.Maskf(incompatibleUnderlyingResourceError, "expected %T", underlying)
```

Comment:
The error message is wrong. We expected `*originalframework.CRUDResource`. We got `underlying`.

Reply:
`underlying` is of type `*originalframework.CRUDResource`.

https://play.golang.org/p/aV_Co3kCMbf

Reply:
I see, `underlying` is always the asserted type. That we even have to talk about this shows again that it is just better to be explicit, instead of smart. Lets maybe just place the type explicitly.

Reply:
It isn't about being smart. It's practical. When the type change the error message is autoupdated. You don't have to keep it in sync which is difficult.

Reply:
Which is implicit, and not explicit, which makes it harder to brain.